adlfs/spec.py

```diff
@@ -1593,17 +1593,26 @@ async def _expand_path(
         return list(sorted(out))
 
     async def _put_file(
-        self, lpath, rpath, delimiter="/", overwrite=False, callback=None, **kwargws
+        self, lpath, rpath, delimiter="/", overwrite=None, callback=None, **kwargs
     ):
         """
         Copy single file to remote
 
         :param lpath: Path to local file
         :param rpath: Path to remote file
         :param delimitier: Filepath delimiter
-        :param overwrite: Boolean (False).  If True, overwrite the existing file present
+        :param overwrite: Boolean (True when filesystem is version aware, otherwise
+            False). If True, overwrite the existing file present
         """
 
+        if overwrite is None:
+            overwrite = self.version_aware
```

Comment:
Can you explain more about why you think the default behavior should depend on whether the filesystem is version aware? Is it just that you can revert the change in a version-aware filesystem?

IMO, I'd rather have consistent behavior at the library level. I don't think I would want this operation to depend on how the filesystem is configured.

Reply:
I agree it should be consistent and that it shouldn't depend on the filesystem, but I don't think the existing `overwrite=False` default makes sense to have at all for the version-aware use case.

So if the preference is to keep the existing default for non-versioned filesystems in the short term, I think there should be separate behavior for the version-aware use case until the future change is made to make the default `overwrite=True` in all case.

Reply:
Can you check the other fsspec implementations? I believe that s3 / s3fs should have versioned filesystems. I'd prefer to match that behavior.

Reply:
s3fs and gcsfs don't have the `overwrite` parameter, and both implementations always overwrite on `put_file()` regardless of whether or not the fs is versioned

Reply:
I agree that making `overwrite` default value dependant on `version_aware` feels really odd. Feels like for the sake of consistency it is better to switch overwrite default to True and then we won't need to do anything else here, right?

Reply:
Just seems considerate to people who are relying on the current behavior.

But... now that I think of it, users of adlfs through higher-level systems like dask and pandas won't actually have a chance to set `overwrite=True`. It would only be helpful if you're using it directly (or if we jumped through a bunch of hoops and made it a setting on the filesystem too, which sounds too complicated).

So I guess I'm OK with just changing the default to match other the libraries.